module/zfs/vdev_anyraid.c

```diff
@@ -0,0 +1,1451 @@
+	uint64_t size = logical_rs->rs_end - logical_rs->rs_start;
+
+	physical_rs->rs_start = child_offset;
+	physical_rs->rs_end = child_offset + size;
+	remain_rs->rs_start = 0;
+	remain_rs->rs_end = 0;
+}
+
+static uint64_t
+vdev_anyraid_nparity(vdev_t *vd)
+{
+	vdev_anyraid_t *var = vd->vdev_tsd;
+	return (var->vd_nparity);
+}
+
+static uint64_t
+vdev_anyraid_ndisks(vdev_t *vd)
+{
+	return (vd->vdev_children);
+}
+
+/*
+ * Functions related to syncing out the tile map each TXG.
+ */
+static boolean_t
+map_write_loc_entry(anyraid_tile_node_t *arn, void *buf, uint32_t *offset)
+{
+	anyraid_map_loc_entry_t *entry = (void *)((char *)buf + *offset);
+	entry->amle_type = AMET_LOC;
+	entry->amle_disk = arn->atn_disk;
+	entry->amle_offset = arn->atn_offset;
+	*offset += sizeof (*entry);
+	return (*offset == SPA_MAXBLOCKSIZE);
+}
+
+static boolean_t
+map_write_skip_entry(uint32_t tile, void *buf, uint32_t *offset,
+    uint32_t prev_id)
+{
+	anyraid_map_skip_entry_t *entry = (void *)((char *)buf + *offset);
+	amse_set_type(entry);
+	amse_set_tile_id(entry, tile - prev_id - 1);
+	*offset += sizeof (*entry);
+	return (*offset == SPA_MAXBLOCKSIZE);
+}
+
+static void
+anyraid_map_write_done(zio_t *zio)
+{
+	abd_free(zio->io_abd);
+}
+
+static void
+map_write_issue(zio_t *zio, vdev_t *vd, uint64_t base_offset, uint8_t idx,
+    abd_t *abd, zio_eck_t *cksum_out, int flags)
+{
+	zio_nowait(zio_write_phys(zio, vd, base_offset +
+	    idx * VDEV_ANYRAID_MAP_SIZE +
+	    VDEV_ANYRAID_MAP_HEADER_SIZE(vd->vdev_ashift), SPA_MAXBLOCKSIZE,
+	    abd, ZIO_CHECKSUM_ANYRAID_MAP, anyraid_map_write_done, cksum_out,
+	    ZIO_PRIORITY_SYNC_WRITE, flags, B_FALSE));
+}
+
+static void
+vdev_anyraid_write_map_done(zio_t *zio)
+{
+	uint64_t *good_writes = zio->io_private;
+
+	if (zio->io_error == 0 && good_writes != NULL)
+		atomic_inc_64(good_writes);
+}
+
+void
+vdev_anyraid_write_map_sync(vdev_t *vd, zio_t *pio, uint64_t txg,
+    uint64_t *good_writes, int flags, vdev_config_sync_status_t status)
+{
+	vdev_t *anyraidvd = vd->vdev_parent;
+	ASSERT3P(anyraidvd->vdev_ops, ==, &vdev_anyraid_ops);
+	spa_t *spa = vd->vdev_spa;
+	vdev_anyraid_t *var = anyraidvd->vdev_tsd;
+	uint32_t header_size = VDEV_ANYRAID_MAP_HEADER_SIZE(vd->vdev_ashift);
+	uint32_t full_size = VDEV_ANYRAID_SINGLE_MAP_SIZE(vd->vdev_ashift);
+	uint32_t nvl_bytes = VDEV_ANYRAID_NVL_BYTES(vd->vdev_ashift);
+	uint8_t update_target = txg % VDEV_ANYRAID_MAP_COPIES;
+	uint64_t base_offset = VDEV_LABEL_START_SIZE +
+	    update_target * full_size;
+
+	abd_t *header_abd =
+	    abd_alloc_linear(header_size, B_TRUE);
+	abd_zero(header_abd, header_size);
+	void *header_buf = abd_borrow_buf(header_abd, header_size);
+	zio_eck_t *cksums = (zio_eck_t *)&((char *)header_buf)[nvl_bytes];
+
+	abd_t *map_abd = abd_alloc_linear(SPA_MAXBLOCKSIZE, B_TRUE);
+	uint8_t written = 0;
+	void *buf = abd_borrow_buf(map_abd, SPA_MAXBLOCKSIZE);
+
+	rw_enter(&var->vd_lock, RW_READER);
+	anyraid_tile_t *cur = avl_first(&var->vd_tile_map);
```

Comment:
As I understand, you have only one copy of a tile map, that covers all TXGs.  It may be OK as long as you don't need precise accounting and not going to ever free the tiles.  But what is not OK, I suppose, is that for writes done in open context (such as Direct I/O and ZIL) maps will not be written till the end of that TXG.  Direct I/O might not care, but it may be impossible to replay the ZIL after crash if its blocks or blocks they reference were written to a new tiles that are not yet synced.